features/environment.py

```diff
@@ -0,0 +1,69 @@
+import datetime
+import subprocess
+
+from behave.runner import Context
+
+from features.util import launch_lxd_container, lxc_exec
+
+
+def before_all(context: Context) -> None:
+    """behave will invoke this before anything else happens.
+
+    In this function, we launch a container, install ubuntu-advantage-tools and
+    then capture an image.  This image is then reused by each feature, reducing
+    test execution time.
+    """
+    create_trusty_uat_lxd_image(context)
+
+
+def _capture_image(container_name: str, image_name: str) -> None:
+    """Capture a lxd container as an image.
+
+    :param container_name:
+        The name of the container to be captured.  Note that this container
+        will be stopped.
+    :param image_name:
+        The name under which the image should be published.
+    """
+    subprocess.run(["lxc", "stop", container_name])
+    subprocess.run(["lxc", "publish", container_name, "--alias", image_name])
+
+
+def create_trusty_uat_lxd_image(context: Context) -> None:
+    """Create a trusty lxd image with ubuntu-advantage-tools installed
+
+    This will launch a container, install ubuntu-advantage-tools, and publish
+    the image.  The image's name is stored in context.image_name for use within
+    step code.
+
+    :param context:
+        A `behave.runner.Context`; this will have `image_name` set on it.
+    """
+    now = datetime.datetime.now()
+    context.image_name = "behave-image-build-" + now.strftime("%s%f")
+
+    launch_lxd_container(context, "ubuntu:trusty", context.image_name)
+    _install_uat_in_container(context.image_name)
+    _capture_image(context.image_name, context.image_name)
```

Comment:
We probably either need to add a context.add_cleanup(cleanup_image) here to [lxc image delete context.image_name]  or pass context into _capture_image and add cleanup in there so we don't start leaving stale test images in lxc.

Reply:
Yes, good catch, thanks!

I would like to do some image reuse (so that if you're iterating on tests locally you don't have to build a new image for every single test run), but we can still do cleanup properly in the interim. I'll file an issue for the reuse.

Reply:
I opened https://github.com/CanonicalLtd/ubuntu-advantage-client/issues/897 for the follow-up I mentioned.